File: selected_coding_interview/codes/python/lc_371_sum_of_two_integers.py
"""
File: lc_371_sum_of_two_integers.py
Created Time: 2025-12-30
Author: krahets
"""

from include import *


# ===== Solution Code =====
class Solution:
    def getSum(self, a: int, b: int) -> int:
        x = 0xffffffff
        a, b = a & x, b & x
        # 循环，当进位为 0 时跳出
        while b != 0:
            # a, b = 非进位和, 进位
            a, b = (a ^ b), (a & b) << 1 & x
        return a if a <= 0x7fffffff else ~(a ^ x)


# ======= Test Case =======
# Test case 1
# TODO: Add appropriate test case

# ====== Driver Code ======
slt = Solution()
# result = slt.getSum(...)
# print(result)

File: selected_coding_interview/codes/java/lc_371_sum_of_two_integers/lc_371_sum_of_two_integers.java
/*
* File: lc_371_sum_of_two_integers.java
* Created Time: 2025-12-30
* Author: krahets
*/

package lc_371_sum_of_two_integers;

import include.*;
import java.util.*;

// ===== Solution Code =====
    class Solution {
        public int getSum(int a, int b) {
            // 循环，当进位为 0 时跳出
            while (b != 0) {
                int c = (a & b) << 1;  // c = 进位
                a ^= b; // a = 非进位和
                b = c; // b = 进位
            }
            return a;
        }
    }

public class lc_371_sum_of_two_integers {
    public static void main(String[] args) {
        // ======= Test Case =======
        // Test case 1
        // TODO: Add appropriate test case

        // ====== Driver Code ======
        Solution slt = new Solution();
        // result = slt.getSum(...)
        // print(result)

    }
}
